### DISCUSSION TITLE
How is the Glfw transparent window

### DISCUSSION THREAD

**Discussion by GeneralBombe** [coderedart/egui_overlay#36] [Q&A] [answered]
Hello, 
i am struggling to get a transparent glfw window and use glow with it. This crate is the closest to what i am trying to achieve, the transparent window actually works. All my attempts simply render the transparent pixels as black. 
I don't know if this is the right place to ask, but is it possible to render opengl itself with this crate, not egui (or maybe additional). 
Alternativley could anyone maybe help me in understanding what the difference is between the window initialisation (and drawing transparency) in this crate and my own project is? 
This is how i try to do this myself, but it does not work and renders black instead of transparent: 
```rust
extern crate glfw;
use glfw::{Action, Context, Key, WindowHint};
use glow::*;

fn main() {
    let mut glfw = glfw::init(glfw::fail_on_errors).unwrap();
    // Enable alpha channel for the window
    glfw.window_hint(WindowHint::TransparentFramebuffer(true));
    glfw.window_hint(WindowHint::Maximized(false));
    glfw.window_hint(WindowHint::ContextVersionMajor(3));
    glfw.window_hint(WindowHint::ContextVersionMinor(3));
    glfw.window_hint(WindowHint::OpenGlProfile(glfw::OpenGlProfileHint::Core));
    
    

    let (mut window, events) = glfw.with_primary_monitor(|glfw, m| {
        glfw.create_window(200, 200, "Window",
            m.map_or(glfw::WindowMode::Windowed, |m| glfw::WindowMode::Windowed))
    }).expect("Failed to create GLFW window");

    window.set_key_polling(true);
    window.make_current();
    let gl = unsafe {
        glow::Context::from_loader_function(|s| window.get_proc_address(s) as *const _)
    };
    
    while !window.should_close() {
        unsafe { 
            gl.enable(glow::BLEND);
            gl.clear_color(0.0, 0.0, 0.0, 0.0);
            
            gl.clear(glow::COLOR_BUFFER_BIT);
        }
        
        glfw.poll_events();
        
        for (_, event) in glfw::flush_messages(&events) {
            handle_window_event(&mut window, event);
        }
        
        window.swap_buffers();
    }
}

fn handle_window_event(window: &mut glfw::Window, event: glfw::WindowEvent) {
    match event {
        glfw::WindowEvent::Key(Key::Escape, _, Action::Press, _) => {
            window.set_should_close(true)
        }
        _ => {}
    }
}
```
Render Doc even shows me, that the pixels should be transparent (if i understand it correctly): 
![grafik](https://github.com/coderedart/egui_overlay/assets/72839268/23715e68-7ef0-4e2d-aa9b-a6e55b1e1be9)

Or does this crate not render with glow at all? I thought egui_render_three_d uses egui_render_glow which uses glow, but maybe i am understanding the code wrong.  
I really hope its appropriate for me to ask this question here.
I would really appreciate any help and thank you in advance :)

**Comment by coderedart** [comment, marked as answer]
You are doing everything correct. When I copy paste your example code into an empty project, transparency works fine for me. 
![image](https://github.com/coderedart/egui_overlay/assets/24411704/3665bfa7-18c2-4ade-bdd5-5c4fb0722c00)

If its any help, you can look at the glfw code at https://github.com/coderedart/egui_overlay/blob/master/crates/egui_window_glfw_passthrough/src/lib.rs#L122
The only differences are:
1. I don't set any context version or opengl profile hints.
4. I don't use primary monitor when creating a window
5. I specifically set opengl API hint (vulkan transparency is way more messy than openGL)


Its hard to pinpoint any issues without knowing your platform details. 
For Microsoft Windows, there's often issues with drivers. Another similar overlay has a decent troubleshooting guide for black transparent windows at https://blishhud.com/docs/user/troubleshooting#1-i-see-the-blish-hud-icons-on-a-black-background . Try checking if it fixes any issues

**Comment by GeneralBombe** [comment]
I really appreciate the answer and the links! Ill try to play around with everything for a bit, maybe i'll find a solution. 
I am using a AMD GPU, but i mean it kinda works, just only one one window. Ill try it without primary window tomorrow, thank you again!
